### DISCUSSION TITLE
Converge and freeze coarse AMR level

### DISCUSSION THREAD

**Discussion by drjosephbrock** [AMReX-Codes/amrex#3964] [Ideas]
Is it possible (and/or simple) to run a simulation (in my case an Euler simulation) on a coarse mesh with several AMR levels defined. I am able to only activate the next level when my current level has reached some global residual reduction. I would like to only iterate on the current level and bypass the coarser level. In other words, I would like to refine the mesh as the solution converges and only spend time solving on the current level while freezing the previous coarse state. I am currently interested in steady-state problems, and I think this would help accelerate my simulations greatly. Any information and advice would be greatly appreciated. Thanks!

**Comment by asalmgren** [comment]
You can certainly do this by inserting some code that controls when a
particular level advances the solution.   But you'll need to give us more
information about what code you're using and how you are handling the
calling of finer levels.

For example, are you using an AmrCore or AmrLevel approach?  Or have you
built your own?

On Wed, May 29, 2024 at 12:22 PM drjosephbrock ***@***.***>
wrote:

> Is it possible (and/or simple) to run a simulation (in my case an Euler
> simulation) on a coarse mesh with several AMR levels defined. I am able to
> only activate the next level when my current level has reached some global
> residual reduction. I would like to only iterate on the current level and
> bypass the coarser level. In other words, I would like to refine the mesh
> as the solution converges and only spend time solving on the current level
> while freezing the previous coarse state. I am currently interested in
> steady-state problems, and I think this would help accelerate my
> simulations greatly. Any information and advice would be greatly
> appreciated. Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AMReX-Codes/amrex/discussions/3964>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRE6YTXXAT6MFPWZ6HTTVTZEYTF5AVCNFSM6AAAAABIPRTJ6SVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZWG42DQOBUGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Ann Almgren
Senior Scientist; Dept. Head, Applied Mathematics
Pronouns: she/her/hers

**Reply by drjosephbrock** [reply]
I am using AmrLevel for my approach, with my solver resembling he CNS example code from the amrex-tutorial repo.  I see that using AmrCore would have direct control over the "evolve" function, whereas the AmrLevel does not. How difficult would it be to augment my approach to control level advancing? Does it make sense to convert to AmrCore?

**Reply by WeiqunZhang** [reply]
You can have a class derived from Amr and override `virtual void coarseTimeStep (Real stop_time);`. Then you have full control.

**Comment by asalmgren** [comment]
You could also do something like copy the old data into the new data
instead of evolving it - and just wrap that in an if test on level or time
or both…

Ann Almgren
Senior Scientist; Dept. Head, Applied Mathematics
Pronouns: she/her/hers
